projects/plutus/src/ltc_afe.c

```diff
@@ -0,0 +1,309 @@
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec & 0xFF);
+
+  return spi_exchange(afe->spi_port, cmd, 4, data, len);
+}
+
+static StatusCode prv_read_voltage(LTCAFESettings *afe, LTCAFEVoltageRegister reg, uint8_t *data) {
+  if (reg > NUM_LTC_AFE_VOLTAGE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // 6 bytes in register + 2 bytes for PEC
+  size_t len = ((LTC6804_CELLS_IN_REG * 2) + 2) * LTC_AFE_DEVICES_IN_CHAIN;
+
+  return prv_read_register(afe, s_voltage_reg[reg], data, len);
+}
+
+// start cell voltage conversion
+static void prv_trigger_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADCV command
+  uint16_t adcv = LTC6804_ADCV_RESERVED | LTC6804_ADCV_DISCHARGE_PERMITTED
+                  | LTC6804_CNVT_CELL_ALL | (mode << 7);
+  uint8_t cmd[4] = { 0 };
+
+  cmd[0] = (uint8_t)(adcv >> 8);
+  cmd[1] = (uint8_t)(adcv & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec);
+
+  prv_wakeup_idle(afe);
+
+  spi_exchange(afe->spi_port, cmd, 4, NULL, 0);
+
+  // wait for conversions to finish
+  delay_ms(100);
+}
+
+static void prv_trigger_aux_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADAX
+  uint16_t adax = LTC6804_ADAX_RESERVED | LTC6804_ADAX_GPIO1 | (mode << 7);
+  uint8_t cmd[4] = { 0 };
+  cmd[0] = (uint8_t)(adax >> 8);
+  cmd[1] = (uint8_t)(adax & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec & 0xFF);
+
+  spi_exchange(afe->spi_port, cmd, 4, NULL, 0);
+
+  // wait for conversions to finish
+  delay_ms(10);
+}
+
+// write config to all devices
+static StatusCode prv_write_config(const LTCAFESettings *afe, uint8_t gpio_enable_pins) {
+  // see p.54 in datasheet
+  // (2 bits for WRCFG + 2 bits for WRCFG PEC) +
+  // (6 bits for CFGR + 2 bits for CFGR PEC) * LTC_AFE_DEVICES_IN_CHAIN
+  uint8_t configuration_cmd[(2 + 2) + (6 + 2) * LTC_AFE_DEVICES_IN_CHAIN] = { 0 };
+  uint8_t cfg_len = (2 + 2) + ((6 + 2) * LTC_AFE_DEVICES_IN_CHAIN);
+  uint8_t configuration_index = 0;
+
+  // WRCFG
+  uint16_t wrcfg = LTC6804_WRCFG_RESERVED;
+  configuration_cmd[configuration_index + 0] = (uint8_t)(wrcfg >> 8);
+  configuration_cmd[configuration_index + 1] = (uint8_t)(wrcfg & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(configuration_cmd, 2);
+  configuration_cmd[configuration_index + 2] = (uint8_t)(cmd_pec >> 8);
+  configuration_cmd[configuration_index + 3] = (uint8_t)(cmd_pec & 0xFF);
+
+  configuration_index += 4;
+  // send CFGR registers starting with the bottom slave in the stack
+  for (uint8_t device = LTC_AFE_DEVICES_IN_CHAIN; device > 0; --device) {
+    uint8_t current_device = LTC_AFE_DEVICES_IN_CHAIN - device;
+    uint8_t enable = gpio_enable_pins;
+    uint16_t undervoltage = 0;
```

Comment:
I wasn't planning on using it, since we're reading all the voltages anyways. However, it's a matter of writing to the config registers, and then checking the status registers, if we decide to do so.